Modifier keys fix

auibrian opened the pull request:
This PR addresses issue #29 
This adds an event tap to the daemon's runloop so that whenever the mouse moves we check the modifiers on the keyboard and make sure they are added to the event rather than lost. This rectifies the loss of modifier keys for things like cmd-click and alt-drag etc.

auibrian committed:
fix cflags being wiped, was breaking mac build

auibrian committed:
add a cgeventtap to fix loss of modifier keys

This adds an event tap to the daemon's runloop so that whenever the mouse moves we check the modifiers on the keyboard and make sure they are added to the event rather than lost. This rectifies the loss of modifier keys for things like cmd-click and alt-drag etc.

light2yellow commented on src/ckb-daemon/usb_mac.c:
```diff
@@ -271,6 +271,67 @@ static void pipecomplete(void* refcon, IOReturn result, void* arg0){
+                usbdevice *kb = keyboard + i;
+                if (!IS_MOUSE_DEV(kb)) {
+                    pthread_mutex_lock(imutex(kb));
+                    CGEventSetFlags(event, (kCGEventFlagMaskNonCoalesced | kb->modifiers));
```
Yep, that's it, the `NonCoalesced` flag that was missing.

fleischie commented:
@auibrian Thank you very much for the PR.

Could you add some comments, on what each segment of code exactly does. This will be very helpful for maintainability reasons. But this is just a friendly suggestion. :smile:

fleischie commented on src/ckb-daemon/usb_mac.c:
```diff
@@ -271,6 +271,67 @@ static void pipecomplete(void* refcon, IOReturn result, void* arg0){
     (*handle)->ReadPipeAsync(handle, ctx->pipe, buffer, ctx->maxsize, pipecomplete, ctx);
 }
 
+CFMachPortRef event_tap;
```
This declaration should be on top of the file. Also there is a documentation missing: What does this reference store?

Also, is it a static variable inside this file?

frickler24 replied:
It must be a static variable, because no other .c file is changed yet.
Pease add it with comments in the usb.h (usb.h ist the header file for both usb_{llinux,mac}.c

fleischie commented on src/ckb-daemon/usb_mac.c:
```diff
@@ -271,6 +271,67 @@ static void pipecomplete(void* refcon, IOReturn result, void* arg0){
 
+CFMachPortRef event_tap;
+
+CGEventRef cg_callback(CGEventTapProxy proxy, CGEventType type, CGEventRef event, void* refcon) {
```
What does this function do exactly? Why is it called cg_callback? Can't there be no other CoreGraphics-related callbacks inside this function?

fleischie commented on src/ckb-daemon/usb_mac.c:
```diff
@@ -271,6 +271,67 @@ static void pipecomplete(void* refcon, IOReturn result, void* arg0){
+                    pthread_mutex_lock(imutex(kb));
+                    CGEventSetFlags(event, (kCGEventFlagMaskNonCoalesced | kb->modifiers));
+                    pthread_mutex_unlock(imutex(kb));
+                    break;
```
It seems to me, that there is the following assumption: There is at most 1 non-mouse keyboard device. Otherwise I don't understand the `break` statement after there are certain flags set for the given event.

Especially here some comments might come in handy.

fleischie commented on src/ckb-daemon/usb_mac.c:
```diff
@@ -271,6 +271,67 @@ static void pipecomplete(void* refcon, IOReturn result, void* arg0){
+}
+
+void registerCGEventTap(CFRunLoopRef run_loop) {
+
```
This empty line is unnecessary.

fleischie commented on src/ckb-daemon/usb_mac.c:
```diff
@@ -271,6 +271,67 @@ static void pipecomplete(void* refcon, IOReturn result, void* arg0){
+    CGEventMaskBit(kCGEventTabletProximity) |
+    CGEventMaskBit(kCGEventOtherMouseDown) |
+    CGEventMaskBit(kCGEventOtherMouseUp) |
+    CGEventMaskBit(kCGEventOtherMouseDragged);
```
I would prefer the mask-bits to be indented (either under the top mask-bit or all of them by two additional indents). This is just a suggestion, as it is a bit hard to make out the intention of this code block at a brief look.

fleischie commented on src/ckb-daemon/usb_mac.c:
```diff
@@ -271,6 +271,67 @@ static void pipecomplete(void* refcon, IOReturn result, void* arg0){
+    CGEventMaskBit(kCGEventOtherMouseUp) |
+    CGEventMaskBit(kCGEventOtherMouseDragged);
+    
+
```
This empty line is unnecessary.

fleischie commented on src/ckb-daemon/usb_mac.c:
```diff
@@ -271,6 +271,67 @@ static void pipecomplete(void* refcon, IOReturn result, void* arg0){
+    return event;
+}
+
+void registerCGEventTap(CFRunLoopRef run_loop) {
```
Again a short description, of what this routine does would be much appreciated.

fleischie reviewed (changes requested):
The PR seems to compile.

Although I cannot reproduce the Cmd/Shift + Select misbehavior, I would appreciate another reviewer to verify, that the broken cmd/shift + select is rectified by this changes.

auibrian commented:
Hi @fleischie, duly noted :)

I agree some commenting would be useful as well, I will to address all of this in the evening when I'm off work.

auibrian committed:
add requested PR changes

- Add comments to the code
- Cleanup code, move global variable declaration to top of file, cleanup spacing and indenting
- fixed issue where if an apple keyboard was also plugged in the modifier keys were being lost on mouse events

auibrian commented:
@fleischie I've pushed some changes to address your concerns.

fleischie commented:
@auibrian Looking good. I'll have to check, whether it's still compiling and whether it does what it says, but then I am all for merging.

Any further concerns, @frickler24, @tatokis, @light2yellow?

fleischie commented:
I approve: It compiles and it seems to do what it says it does. I would appreciate it, if somebody could evaluate, whether it compiles and does what it says it does on another machine.

Then I might just merge this PR.

jpagano commented:
Works fine! 😄

tatokis commented:
Before this gets merged, could someone with an earlier version of the OS test and make sure this change doesn't break anything?

ghost commented:
@tasossah very valid concern, Karabiner-Elements have only 10.11.x+ support. Somebody with 10.10.x or lower must test it, if it doesn't work we can add an if clause.

On Mar 9, 2017, 22:38, at 22:38, Tasos Sahanidis <notifications@github.com> wrote:
>Before this gets merged, could someone with an earlier version of the
>OS test and make sure this change doesn't break anything?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/mattanger/ckb-next/pull/106#issuecomment-285491574

light2yellow commented on src/ckb-daemon/usb_mac.c:
```diff
@@ -838,6 +931,8 @@ int usbmain(){
     if(iterator_hid)
         iterate_devices_hid(0, iterator_hid);
 
+    register_mouse_event_tap(mainloop);
```
Actually we don't need to test this on older implementations. We can just wrap it around like this
```c
#if __MAC_OS_X_VERSION_MAX_ALLOWED >= 1012
register_mouse_event_tap(mainloop);
#endif
```
based on [the docs](https://developer.apple.com/library/content/documentation/DeveloperTools/Conceptual/cross_development/Using/using.html) and everything should be fine.

What do you think?

fleischie replied:
This is a valid approach, as it seems this fix only/mostly targets Sierra. I am for Sierra-guarding. 😄

auibrian replied:
Works for me. I can do that this evening.

light2yellow replied:
@tatokis made an important point that this is a preprocessor if clause, which means we would need to provide more than one standalone .pkg.

philmb3487 commented:
Yep, that works for me. STRAFE on Sierra. Thank you

ghost commented:
# 10.10.x & 10.11.x users, please test
### 10.9.x and below would also be useful, but not crucial

mik115 commented:
This works fine on 10.12.3 with K65

linjer commented:
Thanks for this change!

I was able to test and it works on 10.12.3 with the K70 RGB

SeNeO commented:
Works fine with a Strafe RGB on 10.12.3.

Edit: There seems to be an issue after a reboot, the modifier doesn't work until I manually restart the CKB daemon.

syropian commented:
Hoping this gets merged soon 🙏

tatokis commented:
@syropian it will, once we find someone to test it on older versions of the OS. (See comment above.)

rkitover commented:
Just tried on sierra with a K70 Lux RGB, can confirm that I can cmd+click links in iterm2 again!

Thank you so much for this! I hope it gets merged soon!

rkitover commented:
If I may make a suggestion, if you want more testers it may make sense put a built `.pkg` file as a github test release (you can upload files to releases) and put a note with a link to it in the README.md.

It seems there is a script in `osx/` to do everything necessary to create a package.

The only thing would be signing the app and the package so that users don't have to disable gatekeeper, someone who is in the apple developer program would have to do that, but disabling gatekeeper is easy so this is not that big a deal.

ghost commented:
For those who would like to test/install but don't have experience with git & GitHub, here's a [short summary](https://github.com/mattanger/ckb-next/issues/29#issuecomment-289412693) for this specific issue.

shippak commented:
Just tested this under macOS Sierra 10.12.4 and it works. Thanks

linjer commented:
I think I've discovered an interesting nuance. I use synergy to share my keyboard/mouse across my macOS and linux workstations.

The host is running macOS Sierra 10.12.4 with the Corsair K70 RGB. When I go to the other machine (RHEL) and use any modifier keys, when I come back to macOS, it's like the modifier keys are stuck until I press one of them once, then it clears.

ghost commented:
@auibrian you said we need to be logged in order to create a CGEventTap. Right now ckb-next is categorized as a _daemon_ (in terms of launchd), but can't we create an global or a user _agent_, additionally (or replacing the daemon, however, I don't know what are the consequences for a multi-user system)? Could this be a solution?

rkitover commented:
@light2yellow, @auibrian if the issue is the daemon launching too early and needing to launch when there is a GUI user, maybe it could just poll for the existance of a logged in console user using something like the tiny code snippet here:

https://developer.apple.com/library/content/qa/qa1133/_index.html

lstkz commented:
@light2yellow 
Regarding #154 
What exactly should I check? Do I need to check out this pull request?

ghost commented:
@lsentkiewicz 
Please install this PR as per https://github.com/mattanger/ckb-next/issues/29#issuecomment-289412693, then reboot and tell us whether you can still use Cmd+click, alt+drag and other combinations with modifier keys (disable Karabiner before the check). This is aimed to check if older, pre-Sierra versions are still working with the new code changes required by Sierra. A supporting issue for this is #29, you might want to read through it (and through this PR as well) to understand better what you should check.

Also, please don't disappear as this PR is not rigorous and will be updated, and when it will be reported to solve all the issues, we will sincerely ask you to test it once again (and it won't be merged until backwards compatibility with at least 10.11 is achieved). Nevertheless, we would like to know about any issues on older macOS versions now for better estimations.

You are the first person with 10.11 so far, therefore very valuable :)

@rkitover 
As the article says, the architecture gets broken indeed (ckb-daemon running as a root daemon starts depending on a GUI with the user rights). In the worst-case scenario we might use this crutch, but I'd rather avoid it and think about other workarounds. One of them is the introduction of an agent in addition to the daemon (however, I know very little about this and it may not be a solution at all) or there might be a possibility to use different functions from the underlying API. Let's wait for @auibrian to tell us what he thinks.

auibrian commented:
Checking for the console user having changed is exactly what I implemented, using the same document from Apple that @rkitover mentions.

I considered using fast user notifications but I didn't see a way to get that outside of Cocoa other than using Carbon which doesn't feel right given it is deprecated.

I looked into agents but by chance I had just spent a bunch of time on the family macbook air deleting user agents that were popping up with errors whenever different users logged into their machines.

As it is the daemon's already violating various things Apple say not to do (`setgid` for instance). Checking the console user doesn't seem to me to be exceedingly heinous.

/cc @light2yellow

tatokis commented:
It is clear to me no one cares enough to test this on older versions. (Or has done it successfully for that matter.)
I suggest we merge it, and if something breaks on older versions, we then deal with it.
In other words, we can't block this forever.

I also agree with @auibrian, and I don't think we are yet capable of rewriting the daemon to work properly on macOS, as per Apple's guidelines.

After this gets merged, then @auibrian can also open another PR with the user context fix.

lstkz commented:
@tatokis
Sorry, I forgot about this issue.

I tested this PR on 10.11 and modifier keys work correctly.

My steps:
- old cbk was installed
- clone the repo
- run `git fetch origin pull/106/head:testing`
- run `./quickinstall`
- restart

tatokis commented:
Spoke too soon then, thank you kindly @lsentkiewicz.
I propose we merge this ASAP.

fleischie committed:
Merge branch 'testing' into issue29

noaht8um commented:
Amazing that you guys fixed this issue!
This has been the reason I stopped using ckb; very glad macOS users can use modifier keys properly now.
Keep up the good work!

noaht8um commented:
Looks like I spoke to soon about it working perfectly haha. Before I installed it, just ran the daemon first (sudo ./ckb-daemon) without using launchctl and then opened the app (open ./ckb.app). It worked great so I installed the launch daemin, but after a restart everything works but the modifier key-clicks. To fix it, I had to manually unload the daemon, wait a minute, and reload it. Obviously this isn't a huge issue, but it's still worth noting I would think. I'd be happy to run any suggested tests to debug this. 

(Not sure if I'm the only one with this issue or not).

ghost commented:
@noaht8um 
Please, read the discussion above. And the appropriate issue this PR closes, too. You're not the only one. However, thanks for describing a temporary solution.

rkitover commented:
@auibrian sorry to bother you, just wanted to clarify something.

You said you are using the console user check previously discussed to check that a GUI user is logged in, does that fix the issue with the daemon not working on restart? (That was the original reason I suggested it.)

The reason I'm asking is that I don't see any of this in this PR's commits list.

lstkz commented:
@tatokis
You ask me to check this issue on 10.11, but now I have this problem on 10.12.5 😞 
This issue occurs after the restart, and I must always reinstall ckb.
Is there any way to fix it?

rkitover commented:
@lsentkiewicz I just put this into my crontab (`crontab -e`):

```crontab
@reboot     sleep 300; sudo pkill ckb-daemon
```

works fine (you need nopassword sudo set up, or install the crontab for root.)

smachi commented:
This pull request fixes this issue:
#190 

But sadly, was not merged yet.
I'm using it since May and still working after every restart (also with 10.12.5 version)

ghost commented:
@smachi 
this PR was merged as you can see. #190 is also a PR, your message makes no sense.

smachi commented:
@light2yellow Why makes no sense?
Perhaps you have not understand me...

I was referring to the #190 PR, not to this PR.
I already tested the merged #106 PR and it's not working after restarting, and the #190 continue working after every restart.
